Add MediaTime::toMicroseconds() and MediaTime::createWithSeconds()

jyavenard opened the pull request:
#### 7f511c9ff686df2fd92e6f9be4228df68ad5541f
<pre>
Add MediaTime::toMicroseconds() and MediaTime::createWithSeconds()
<a href="https://bugs.webkit.org/show_bug.cgi?id=281734">https://bugs.webkit.org/show_bug.cgi?id=281734</a>
<a href="https://rdar.apple.com/138171274">rdar://138171274</a>

Reviewed by Youenn Fablet.

Fly-by: bypass calculation of GCD and LCM for the most common used case: operations of MediaTime where the time scales are equal.
toMicroseconds is made to reduce the chance of overflow when calculating value * 1000000 / scale.

API tests added.

* Source/WTF/wtf/MediaTime.cpp:
(WTF::greatestCommonDivisor):
(WTF::leastCommonMultiple):
(WTF::MediaTime::toMicroseconds const):
(WTF::MediaTime::operator+ const):
(WTF::MediaTime::operator- const):
* Source/WTF/wtf/MediaTime.h:
* Tools/TestWebKitAPI/Tests/WTF/MediaTime.cpp:
(TestWebKitAPI::TEST(WTF, MediaTime)):
(TestWebKitAPI::TEST(WTF, MediaTimeInExpected)):

Canonical link: <a href="https://commits.webkit.org/285597@main">https://commits.webkit.org/285597@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/592a4896cae40927dc2c822bce1551944b6614b6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [⏳ 🧪 style ](https://ews-build.webkit.org/#/builders/Style-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52653 "Build is in progress. Recent messages:Running apply-patch") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75339 "Passed tests") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-2-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-intel-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-18-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-11-Build-EWS "Waiting in queue, processing has not started yet") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WTF/wtf/MediaTime.cpp:
```diff
@@ -155,6 +160,23 @@ double MediaTime::toDouble() const
+int64_t MediaTime::toMicroseconds() const
+{
+    if (isInvalid() || isIndefinite())
+        return std::numeric_limits<int64_t>::quiet_NaN(); // 0
```
Remove // 0?

youennf commented on Source/WTF/wtf/MediaTime.cpp:
```diff
@@ -155,6 +160,23 @@ double MediaTime::toDouble() const
+        return std::numeric_limits<int64_t>::min();
+    if (hasDoubleValue())
+        return m_timeValueAsDouble * 1000000.0;
+    // result = (m_timeValue / m_timeScale) * 1000000 + (m_timeValue % m_timeScale) * 1000000;
```
Remove this as well?

jyavenard replied:
I thought it was useful to have as it's not particularly obvious operation otherwise.

youennf commented on Source/WTF/wtf/MediaTime.cpp:
```diff
@@ -184,11 +206,13 @@ MediaTime MediaTime::operator+(const MediaTime& rhs) const
-    b.setTimeScale(commonTimeScale);
+    uint32_t commonTimeScale = a.m_timeScale;
+    if (a.m_timeScale != b.m_timeScale) {
+        if (!leastCommonMultiple(a.m_timeScale, b.m_timeScale, commonTimeScale) || commonTimeScale > MaximumTimeScale)
```
Why adding the a ==b check here and in leastCommonMultiple?

jyavenard replied:
that made me look at `leastCommonMultiple` and fix a bug there !

youennf commented on Source/WTF/wtf/MediaTime.cpp:
```diff
@@ -228,11 +252,14 @@ MediaTime MediaTime::operator-(const MediaTime& rhs) const
+    uint32_t commonTimeScale = a.m_timeScale;
+    if (a.m_timeScale != b.m_timeScale) {
+        uint32_t commonTimeScale;
+        if (!leastCommonMultiple(this->m_timeScale, rhs.m_timeScale, commonTimeScale) || commonTimeScale > MaximumTimeScale)
```
Ditto.

jyavenard committed:
Add MediaTime::toMicroseconds() and MediaTime::createWithSeconds()
https://bugs.webkit.org/show_bug.cgi?id=281734
rdar://138171274

Reviewed by Youenn Fablet.

Fly-by: bypass calculation of GCD and LCM for the most common used case: operations of MediaTime where the time scales are equal.
toMicroseconds is made to reduce the chance of overflow when calculating value * 1000000 / scale.

API tests added.

* Source/WTF/wtf/MediaTime.cpp:
(WTF::greatestCommonDivisor):
(WTF::leastCommonMultiple):
(WTF::MediaTime::toMicroseconds const):
(WTF::MediaTime::operator+ const):
(WTF::MediaTime::operator- const):
* Source/WTF/wtf/MediaTime.h:
* Tools/TestWebKitAPI/Tests/WTF/MediaTime.cpp:
(TestWebKitAPI::TEST(WTF, MediaTime)):
(TestWebKitAPI::TEST(WTF, MediaTimeInExpected)):

Canonical link: https://commits.webkit.org/285597@main
